### PR TITLE
[BugFix] Do not raise a `ValueError` when `tool_choice` is set to the supported `none` option and `tools` are not defined.

### PR DESCRIPTION
**Pull request by gcalmettes**
This PR fixes the raise of a `ValueError`  when `tool_choice` is set to the supported `none` value.

Currently, the `ChatCompletionRequest` object [supports the `none` and `auto` options](https://github.com/vllm-project/vllm/blob/603a661ae8ccadd8401284f7db8563164b232651/vllm/entrypoints/openai/protocol.py#L178) for the `tool_choice` field.

However, setting `tool_choice` to `none` will raise an error as the current validation check is [expecting the `tools` field to be also set](https://github.com/vllm-project/vllm/blob/603a661ae8ccadd8401284f7db8563164b232651/vllm/entrypoints/openai/protocol.py#L461) when `tool_choice` is defined, whatever its value.

```
{"object":"error","message":"[{'type': 'value_error', 'loc': ('body',), 'msg': 'Value error, When using `tool_choice`, `tools` must be set.', 'input': {'model': 'pixtral', 'messages': [{'role': 'system', 'content': 'You are a helpful assistant'}, {'role': 'user', 'content': 'tell me a short story'}], 'max_tokens': 100, 'temperature': 0.7, 'top_p': 0.7, 'presence_penalty': 0, 'stop': [], 'stream': False, 'tool_choice': 'none'}, 'ctx': {'error': ValueError('When using `tool_choice`, `tools` must be set.')}}]","type":"BadRequestError","param":null,"code":400}%
```

This PR prevents the `ValueError` to occur and also ensure that when `tool_choice` is set to the supported `none` value, no `tools` definitions are actually sent to the model.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by DarkLight1337** [comment]
cc @K-Mistele can you take a look at this and check whether the docs are up-to-date as well?

**Comment by gcalmettes** [comment]
Note that we encountered the issue when using vllm with [Microsoft semantic-kernel](https://github.com/microsoft/semantic-kernel/tree/e79ee3ab0cc91ccc4b295bb429ec7186d14ec770), which forces the `tool_choice` to `none` if no tool_choice is set in the request (see [here](https://github.com/microsoft/semantic-kernel/blob/e79ee3ab0cc91ccc4b295bb429ec7186d14ec770/dotnet/src/Connectors/Connectors.OpenAI/Core/ClientCore.ChatCompletion.cs#L976)).

**Comment by K-Mistele** [comment]
Taking a look; this seems pretty straightforward

File: vllm/entrypoints/openai/protocol.py [hunk 1]
```diff
@@ -454,6 +454,12 @@ def check_tool_usage(cls, data):
         if "tool_choice" not in data and data.get("tools"):
             data["tool_choice"] = "auto"
 
+        # if "tool_choice" is "none" -- ignore tools if present
+        if "tool_choice" in data and data["tool_choice"] == "none":
+            # ensure that no tools are present
+            data.pop("tools", None)
+            return data
+
```
Comment (K-Mistele) [inline vllm/entrypoints/openai/protocol.py:462]: This is a good solution. 

Worth noting, the model could still hallucinate a tool call though (smaller models do this on occasion; I have seem llama and some fine-tunes of it do this), and the tool parser would still parse it and return an assistant-role tool call chat completion.

It should be possible to disable tool parsers for the duration of a request or to bypass them so that even if the model does "try" to call a tool by hallucinating one, it will be returned as an assistant-role text chat completion rather than a tool call chat completion.

This may be overkill. What do you think @gcalmettes @DarkLight1337 @mgoin ?

Comment (K-Mistele) [reply vllm/entrypoints/openai/protocol.py:462]: Otherwise LGTM :)

Comment (DarkLight1337) [reply vllm/entrypoints/openai/protocol.py:462]: > It should be possible to disable tool parsers for the duration of a request or to bypass them so that even if the model does "try" to call a tool by hallucinating one, it will be returned as an assistant-role text chat completion rather than a tool call chat completion.

How would we detect that?

Comment (K-Mistele) [reply vllm/entrypoints/openai/protocol.py:462]: I don't think we need to detect anything, we just disable tool parsing & extraction if the request specifies `tool_choice="none"` in addition to not passing tools to the chat template, so that any tokens generated whether they are text or a hallucinated tool call are processed as normal text due to the tool parser being disabled.

Comment (DarkLight1337) [reply vllm/entrypoints/openai/protocol.py:462]: From my understanding, we should be able to achieve this by updating this code here: https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/serving_engine.py#L473

So that the tool parser isn't even called if `tool_dicts is None`.

Comment (gcalmettes) [reply vllm/entrypoints/openai/protocol.py:462]: @DarkLight1337 @K-Mistele I agree, preventing any tool parsing if `tool_choice` is set to `none` would be nice to prevent hallucinated `tool_calls` to be parsed.
I have added a commit for deactivating the `adjust_request` flow when `tool_choice` is `none` as you suggested  @DarkLight1337

Comment (gcalmettes) [reply vllm/entrypoints/openai/protocol.py:462]: and to bounce back on what you were mentioning @K-Mistele , we are definitely observing the same behavior with Llama 8B (quantized or not), whereas we haven’t seen it yet with Llama 70B or any small Mistral models 

> Worth noting, the model could still hallucinate a tool call though (smaller models do this on occasion; I have seem llama and some fine-tunes of it do this), and the tool parser would still parse it and return an assistant-role tool call chat completion.

Comment (DarkLight1337) [reply vllm/entrypoints/openai/protocol.py:462]: We also need to deactivate calls to `extract_tool_calls` and `extract_tool_calls_streaming`.

Comment (gcalmettes) [reply vllm/entrypoints/openai/protocol.py:462]: @DarkLight1337 from my understanding, the cases of skipping `extract_tool_calls` and `extract_tool_calls_streaming` are already handled:
- for non streaming case, if `tool_choice` is `none` the generated LLM output is [already directly set without tool extraction](https://github.com/vllm-project/vllm/blob/8a4358ecb5ba457fad2be0ed930132489eddddf5/vllm/entrypoints/openai/serving_chat.py#L636) (so the tool extraction process, [here](https://github.com/vllm-project/vllm/blob/8a4358ecb5ba457fad2be0ed930132489eddddf5/vllm/entrypoints/openai/serving_chat.py#L652), is not run in this case)
- similarly in streaming, [this check](https://github.com/vllm-project/vllm/blob/8a4358ecb5ba457fad2be0ed930132489eddddf5/vllm/entrypoints/openai/serving_chat.py#L412) does not match `tool_calls="none"`

Comment (DarkLight1337) [reply vllm/entrypoints/openai/protocol.py:462]: Ah, I see. Should be good then.

File: docs/source/serving/openai_compatible_server.md [hunk 2]
```diff
@@ -189,8 +189,7 @@ The order of priorities is `command line > config file values > defaults`.
 ---
 
 ## Tool calling in the chat completion API
-
-vLLM supports named function calling and `auto` tool choice  in the chat completion API. The `tool_choice` options `required` is **not yet supported** but on the roadmap.
+vLLM supports only named function calling in the chat completion API. The `tool_choice` option `required` is **not yet supported** but on the roadmap.
```
Comment (DarkLight1337) [inline docs/source/serving/openai_compatible_server.md:192]: Is it intentional to remove the mention of `auto` tool choice? This change gives me the impression that it's not supported.

Comment (gcalmettes) [reply docs/source/serving/openai_compatible_server.md:192]: Yes it is intentional. The original sentence was referring to the `auto` and `required` **NOT** being supported, while the `auto` tool choice option is actually supported.

Comment (gcalmettes) [reply docs/source/serving/openai_compatible_server.md:192]: Maybe a better change would be to be more explicit and actually mention that the `none` and `auto` tool_choices options are supported while the `required` option is not yet supported and in the roadmap ?

Comment (gcalmettes) [reply docs/source/serving/openai_compatible_server.md:192]: ha yes, I see what you mean, because I haven't also removed the `only` it doesn't give a very clear picture of what actually vLLM supports or not. I'll update

Comment (gcalmettes) [reply docs/source/serving/openai_compatible_server.md]: Updated !

**Review by DarkLight1337 — APPROVED** [review]
Thanks for your patience!

**Comment by gcalmettes** [comment]
@DarkLight1337 it looks like one of the required CI pipeline (for auto-merge) has failed due to timeout to fetch one of the model. Anything I can do on my end for this ? Or does this required force merge ?

**Comment by DarkLight1337** [comment]
Thanks for the reminder, I have retried the tests just now.
